File: src/reuser/agent/ReuserAgent.php
<?php
/*
 SPDX-FileCopyrightText: © 2014-2018 Siemens AG
Author: Daniele Fognini, Andreas Würl

 SPDX-License-Identifier: GPL-2.0-only
*/

namespace Fossology\Reuser;

use Fossology\Lib\Agent\Agent;
use Fossology\Lib\BusinessRules\AgentLicenseEventProcessor;
use Fossology\Lib\BusinessRules\ClearingDecisionFilter;
use Fossology\Lib\BusinessRules\ClearingDecisionProcessor;
use Fossology\Lib\BusinessRules\ClearingEventProcessor;
use Fossology\Lib\Dao\ClearingDao;
use Fossology\Lib\Dao\UploadDao;
use Fossology\Lib\Dao\AgentDao;
use Fossology\Lib\Proxy\ScanJobProxy;
use Fossology\Lib\Dao\CopyrightDao;
use Fossology\Lib\Data\ClearingDecision;
use Fossology\Lib\Data\DecisionTypes;
use Fossology\Lib\Data\Tree\Item;
use Fossology\Lib\Util\ArrayOperation;
use Fossology\Lib\Data\Clearing\ClearingEventTypes;
use Fossology\Lib\Data\Tree\ItemTreeBounds;

include_once(__DIR__ . "/version.php");

/**
 * @file
 * @brief Reuser agent source
 * @class ReuserAgent
 * @brief The reuser agent
 */
class ReuserAgent extends Agent
{
  /** @var UploadDao $uploadDao
   * UploadDao object
   */
  private $uploadDao;
  /** @var ClearingEventProcessor $clearingEventProcessor
   * ClearingEventProcessor object
   */
  private $clearingEventProcessor;
  /** @var AgentLicenseEventProcessor $agentLicenseEventProcessor
   * AgentLicenseEventProcessor object
   */
  private $agentLicenseEventProcessor;
  /** @var ClearingDecisionFilter $clearingDecisionFilter
   * ClearingDecisionFilter object
   */
  private $clearingDecisionFilter;
  /** @var ClearingDecisionProcessor $clearingDecisionProcessor
   * ClearingDecisionProcessor object
   */
  private $clearingDecisionProcessor;
  /** @var ClearingDao $clearingDao
   * ClearingDao object
   */
  private $clearingDao;
  /** @var CopyrightDao $copyrightDao
   * CopyrightDao object
   */
  private $copyrightDao;
  /** @var AgentDao $aagentDao
   * AgentDao object
   */
  private $aagentDao;
  /** @var DecisionTypes $decisionTypes
   * DecisionTypes object
   */
  private $decisionTypes;

  function __construct()
  {
    parent::__construct(REUSER_AGENT_NAME, AGENT_VERSION, AGENT_REV);
    $this->aagentDao = $this->container->get('dao.agent');
    $this->uploadDao = $this->container->get('dao.upload');
    $this->clearingDao = $this->container->get('dao.clearing');
    $this->copyrightDao = $this->container->get('dao.copyright');
    $this->decisionTypes = $this->container->get('decision.types');
    $this->clearingEventProcessor = $this->container->get('businessrules.clearing_event_processor');
    $this->clearingDecisionFilter = $this->container->get('businessrules.clearing_decision_filter');
    $this->clearingDecisionProcessor = $this->container->get('businessrules.clearing_decision_processor');
    $this->agentLicenseEventProcessor = $this->container->get('businessrules.agent_license_event_processor');
  }

  /**
   * @brief Get the upload items and reuse based on reuse mode
   * @param int $uploadId Upload id to process
   * @see Fossology::Lib::Agent::Agent::processUploadId()
   */
  function processUploadId($uploadId)
  {
    $itemTreeBounds = $this->uploadDao->getParentItemBounds($uploadId);
    foreach ($this->uploadDao->getReusedUpload($uploadId, $this->groupId) as $reuseTriple) {
      // Get the reuse upload id
      $reusedUploadId = $reuseTriple['reused_upload_fk'];
      // Get the group id
      $reusedGroupId = $reuseTriple['reused_group_fk'];
      // Get the reuse mode
      $reuseMode = $reuseTriple['reuse_mode'];
      // Get the ItemTreeBounds for the upload
      $itemTreeBoundsReused = $this->uploadDao->getParentItemBounds($reusedUploadId);
      if (false === $itemTreeBoundsReused) {
        continue;
      }

      if ($reuseMode & UploadDao::REUSE_ENHANCED) {
        $this->processEnhancedUploadReuse($itemTreeBounds, $itemTreeBoundsReused, $reusedGroupId);
      } else {
        $this->processUploadReuse($itemTreeBounds, $itemTreeBoundsReused, $reusedGroupId);
      }

      if ($reuseMode & UploadDao::REUSE_MAIN) {
        $this->reuseMainLicense($uploadId, $this->groupId, $reusedUploadId, $reusedGroupId);
      }

      if ($reuseMode & UploadDao::REUSE_CONF) {
        $this->reuseConfSettings($uploadId, $reusedUploadId);
      }

      if ($reuseMode & UploadDao::REUSE_COPYRIGHT) {
        $this->reuseCopyrights($uploadId, $reusedUploadId);
      }
    }
    return true;
  }
  protected function reuseConfSettings($uploadId, $reusedUploadId)
  {
    if (!$this->uploadDao->insertReportConfReuse($uploadId, $reusedUploadId)) {
      echo "INFO :: Report configuration for select upload doesn't exists. Unable to copy!!!";
    }
    $this->heartbeat(1);
    return true;
  }

  /**
   * @brief Reuse main license from previous upload
   *
   * Get add the main licenses from previous upload and make them in new upload
   * @param int $uploadId       Current upload
   * @param int $groupId        Current user
   * @param int $reusedUploadId Upload to reuse
   * @param int $reusedGroupId  Group of reused upload
   * @return boolean True once finished
   */
  protected function reuseMainLicense($uploadId, $groupId, $reusedUploadId, $reusedGroupId)
  {
    $mainLicenseIds = $this->clearingDao->getMainLicenseIds($reusedUploadId, $reusedGroupId);
    if (!empty($mainLicenseIds)) {
      foreach ($mainLicenseIds as $mainLicenseId) {
        if (in_array($mainLicenseId, $this->clearingDao->getMainLicenseIds($uploadId, $groupId))) {
          continue;
        } else {
          $this->clearingDao->makeMainLicense($uploadId, $groupId, $mainLicenseId);
        }
      }
    }
    return true;
  }

  /**
   * @brief getAgentId from upload
   *
   * Get add the main licenses from previous upload and make them in new upload
   * @param int $uploadId       uplaodId
   * @return int agentId
   */
  protected function getAgentId($uploadId)
  {
    $agentName = 'copyright';

    /** @var ScanJobProxy $scanJobProxy */
    $scanJobProxy = new ScanJobProxy($this->aagentDao, $uploadId);

    $scanJobProxy->createAgentStatus(array($agentName));
    $selectedScanners = $scanJobProxy->getLatestSuccessfulAgentIds();
    if (!array_key_exists($agentName, $selectedScanners)) {
      return;
    }
    return $selectedScanners[$agentName];
  }


  /**
   * @brief Reuse deactivated Copyrights from previous upload
   *
   * Get add the main licenses from previous upload and make them in new upload
   * @param int $uploadId       Current upload
   * @param int $reusedUploadId Upload to reuse
   * @return boolean True once finished
   */
  protected function reuseCopyrights($uploadId, $reusedUploadId)
  {
    $agentId = $this->getAgentId($uploadId);
    $reusedAgentId = $this->getAgentId($reusedUploadId);
    if ($agentId == null || $reusedAgentId == null) {
      return true;
    }
    $uploadTreeTableName = $this->uploadDao->getUploadtreeTableName($uploadId);
    $extrawhere = ' agent_fk='.$agentId;
    $allCopyrights = $this->copyrightDao->getScannerEntries('copyright',
      $uploadTreeTableName, $uploadId, null, $extrawhere);

    $reusedCopyrights = $this->copyrightDao->getAllEventEntriesForUpload(
      $reusedUploadId, $reusedAgentId);

    if (!empty($reusedCopyrights) && !empty($allCopyrights)) {
      foreach ($reusedCopyrights as $reusedCopyright) {
        foreach ($allCopyrights as $copyrightKey => $copyright) {
          if (strcmp($copyright['hash'], $reusedCopyright['hash']) == 0) {
            if ($this->dbManager->booleanFromDb($reusedCopyright['is_enabled'])) {
              $action = "update";
              $content = $reusedCopyright["contentedited"];
            } else {
              $action = "delete";
              $content = "";
            }
            $hash = $copyright['hash'];
            $item = $this->uploadDao->getItemTreeBounds(intval($copyright['uploadtree_pk']),
                      $uploadTreeTableName);
            $this->copyrightDao->updateTable($item, $hash, $content,
              $this->userId, 'copyright', $action);
            unset($allCopyrights[$copyrightKey]);
            $this->heartbeat(1);
          }
        }
      }
    }
    return true;
  }

  /**
   * @brief Compare the files from both uploads and copy decisions
   * @param ItemTreeBounds $itemTreeBounds        Current upload
   * @param ItemTreeBounds $itemTreeBoundsReused  Reused upload
   * @param int $reusedGroupId
   * @throws \Exception
   * @return boolean True once finished
   */
  protected function processUploadReuse($itemTreeBounds, $itemTreeBoundsReused, $reusedGroupId)
  {
    $groupId = $this->groupId;
    $userId = $this->userId;

    $clearingDecisions = $this->clearingDao->getFileClearingsFolder($itemTreeBoundsReused, $reusedGroupId);
    $currenlyVisibleClearingDecisions = $this->clearingDao->getFileClearingsFolder($itemTreeBounds, $groupId);

    $currenlyVisibleClearingDecisionsById = $this->mapByClearingId($currenlyVisibleClearingDecisions);
    $clearingDecisionsById = $this->mapByClearingId($clearingDecisions);

    $clearingDecisionsToImport = array_diff_key($clearingDecisionsById,$currenlyVisibleClearingDecisionsById);

    $clearingDecisionToImportByFileId = $this->mapByFileId($clearingDecisionsToImport);

    $uploadDao = $this->uploadDao;
    /** @var Item[] $containedItems */
    $containedItems = ArrayOperation::callChunked(
        function ($fileIds) use ($itemTreeBounds, $uploadDao)
        {
          return $uploadDao->getContainedItems(
              $itemTreeBounds,
              "pfile_fk = ANY($1)",
              array('{' . implode(', ', $fileIds) . '}')
          );
        }, array_keys($clearingDecisionToImportByFileId), 100);

    foreach ($containedItems as $item) {
      $fileId = $item->getFileId();
      if (empty($fileId)) {
        continue;
      }
      if (array_key_exists($fileId, $clearingDecisionToImportByFileId)) {
        $this->createCopyOfClearingDecision($item->getId(), $userId, $groupId,
          $clearingDecisionToImportByFileId[$fileId]);
      } else {
        throw new \Exception("bad internal state");
      }

      $this->heartbeat(1);
    }

    return true;
  }

  /**
   * @brief Get clearing decisions and use copyClearingDecisionIfDifferenceIsSmall()
   * @param ItemTreeBounds $itemTreeBounds        Current upload
   * @param ItemTreeBounds $itemTreeBoundsReused  Reused upload
   * @param int $reusedGroupId
   * @see copyClearingDecisionIfDifferenceIsSmall()
   */
  protected function processEnhancedUploadReuse($itemTreeBounds, $itemTreeBoundsReused, $reusedGroupId)
  {
    $clearingDecisions = $this->clearingDao->getFileClearingsFolder($itemTreeBoundsReused, $reusedGroupId);
    $currenlyVisibleClearingDecisions = $this->clearingDao->getFileClearingsFolder($itemTreeBounds, $this->groupId);

    $currenlyVisibleClearingDecisionsById = $this->mapByClearingId($currenlyVisibleClearingDecisions);
    $clearingDecisionsById = $this->mapByClearingId($clearingDecisions);

    $clearingDecisionsToImport = array_diff_key($clearingDecisionsById,$currenlyVisibleClearingDecisionsById);

    $sql = "SELECT ut.* FROM uploadtree ur, uploadtree ut WHERE ur.upload_fk=$2"
         . " AND ur.pfile_fk=$3 AND ut.upload_fk=$1 AND ut.ufile_name=ur.ufile_name";
    $stmt = __METHOD__.'.reuseByName';
    $this->dbManager->prepare($stmt, $sql);
    $treeDao = $this->container->get('dao.tree');

    foreach ($clearingDecisionsToImport as $clearingDecision) {
      $reusedPath = $treeDao->getRepoPathOfPfile($clearingDecision->getPfileId());
      if (empty($reusedPath)) {
        // File missing from repo
        continue;
      }

      $res = $this->dbManager->execute($stmt,array($itemTreeBounds->getUploadId(),
        $itemTreeBoundsReused->getUploadId(),$clearingDecision->getPfileId()));
      while ($row = $this->dbManager->fetchArray($res)) {
        $newPath = $treeDao->getRepoPathOfPfile($row['pfile_fk']);
        if (empty($newPath)) {
          // File missing from repo
          continue;
        }
        $this->copyClearingDecisionIfDifferenceIsSmall($reusedPath, $newPath, $clearingDecision, $row['uploadtree_pk']);
      }
      $this->dbManager->freeResult($res);
    }
  }

  /**
   * @brief Use `diff` tool to compare files
   *
   * Uses `diff` tool to compare two files. If the line difference is less
   * than 5, then reuser copy the decisions.
   * @param string $reusedPath
   * @param string $newPath
   * @param ClearingDecision $clearingDecision Array of clearing decisions
   * @param int $itemId
   * @throws \Exception Throws if `diff` tool fails
   */
  protected function copyClearingDecisionIfDifferenceIsSmall($reusedPath,$newPath,$clearingDecision,$itemId)
  {
    $diffLevel = exec("diff $reusedPath $newPath | wc -l");
    if ($diffLevel === false) {
      throw new \Exception('cannot use diff tool');
    }
    if ($diffLevel < 5) {
      $this->createCopyOfClearingDecision($itemId, $this->userId, $this->groupId, $clearingDecision);
      $this->heartbeat(1);
    }
  }

  /**
   * @brief Maps clearing decisions by file id
   *
   * Creates array with file ids as key and ClearingDecision object as value
   * @param ClearingDecision $clearingDecisions  Array of clearing decisions
   * @return ClearingDecision[]
   */
  protected function mapByFileId($clearingDecisions)
  {
    $clearingDecisionByFileId = array();
    foreach ($clearingDecisions as $clearingDecision) {
      $fileId = $clearingDecision->getPfileId();
      if (!array_key_exists($fileId, $clearingDecisionByFileId)) {
        $clearingDecisionByFileId[$fileId] = $clearingDecision;
      }
    }
    return $clearingDecisionByFileId;
  }

  /**
   * @brief Copy clearing decisions from an upload tree item
   * @param int $itemId
   * @param int $userId
   * @param int $groupId
   * @param ClearingDecision $clearingDecisionToCopy
   */
  protected function createCopyOfClearingDecision($itemId, $userId, $groupId, $clearingDecisionToCopy)
  {
    $clearingEventIdsToCopy = array();
    /** @var ClearingEvent $clearingEvent */
    foreach ($clearingDecisionToCopy->getClearingEvents() as $clearingEvent) {
      $licenseId = $clearingEvent->getLicenseId();
      $uploadTreeId = $itemId;
      $isRemoved = $clearingEvent->isRemoved();
      $type = ClearingEventTypes::USER;
      $reportInfo = $clearingEvent->getReportinfo();
      $comment = $clearingEvent->getComment();
      $acknowledgement = $clearingEvent->getAcknowledgement();
      $jobId = $this->jobId;
      $clearingEventIdsToCopy[] = $this->clearingDao->insertClearingEvent(
        $uploadTreeId, $userId, $groupId, $licenseId, $isRemoved,
        $type, $reportInfo, $comment, $acknowledgement, $jobId);
    }

    $this->clearingDao->createDecisionFromEvents(
        $itemId,
        $userId,
        $groupId,
        $clearingDecisionToCopy->getType(),
        $clearingDecisionToCopy->getScope(),
        $clearingEventIdsToCopy
    );
  }

  /**
   * @brief Map clearing decisions by clearing id
   *
   * Creates array with clearing ids as key and ClearingDecision object as
   * value
   * @param ClearingDecision $clearingDecisions  Array of clearing decisions
   * @return ClearingDecision[]
   */
  public function mapByClearingId($clearingDecisions)
  {
    $mapped = array();

    foreach ($clearingDecisions as $clearingDecision) {
      $mapped[$clearingDecision->getClearingId()] = $clearingDecision;
    }

    return $mapped;
  }
}

File: src/reuser/agent/reuser_agent.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <map>
#include <pqxx/pqxx>

// Stub DAO and processor classes for minimal port
class UploadDao {};
namespace UploadDaoConstants {
    constexpr int REUSE_ENHANCED = 1 << 0;
    constexpr int REUSE_MAIN = 1 << 1;
    constexpr int REUSE_CONF = 1 << 2;
    constexpr int REUSE_COPYRIGHT = 1 << 3;
}
class ClearingDao {};
class CopyrightDao {};
class AgentDao {};
class DecisionTypes {};

class ReuserAgent {
public:
    // Member variables (stubs for now)
    std::unique_ptr<UploadDao> uploadDao;
    std::unique_ptr<ClearingDao> clearingDao;
    std::unique_ptr<CopyrightDao> copyrightDao;
    std::unique_ptr<AgentDao> agentDao;
    std::unique_ptr<DecisionTypes> decisionTypes;

    ReuserAgent() {
        // Initialize DAOs (stub)
        uploadDao = std::make_unique<UploadDao>();
        clearingDao = std::make_unique<ClearingDao>();
        copyrightDao = std::make_unique<CopyrightDao>();
        agentDao = std::make_unique<AgentDao>();
        decisionTypes = std::make_unique<DecisionTypes>();
    }

    // --- Ported logic from PHP ---
    // These are stub classes and methods to be implemented
    struct ItemTreeBounds {};
    struct ReuseTriple {
        int reused_upload_fk;
        int reused_group_fk;
        int reuse_mode;
    };
    // Simulate UploadDao constants
    // Use UploadDaoConstants for constants

    // --- Logging helpers ---
    void logInfo(const std::string& msg) {
        std::cerr << "[INFO] " << msg << std::endl;
    }
    void logError(const std::string& msg) {
        std::cerr << "[ERROR] " << msg << std::endl;
    }

    // --- DB connection string from env ---
    std::string getDbConnStr() {
        const char* db = getenv("DB");
        const char* user = getenv("DBUSER");
        const char* pass = getenv("DBPASS");
        const char* host = getenv("DBHOST");
        std::string conn = "dbname=";
        conn += db ? db : "fossology";
        conn += " user=";
        conn += user ? user : "fossy";
        if (pass) { conn += " password="; conn += pass; }
        conn += " host=";
        conn += host ? host : "localhost";
        return conn;
    }

    bool processUploadId(int uploadId) {
        logInfo("Starting processUploadId for upload " + std::to_string(uploadId));
        ItemTreeBounds itemTreeBounds; // = uploadDao->getParentItemBounds(uploadId);
        std::vector<ReuseTriple> reusedUploads = getReusedUploads(uploadId);
        for (const auto& reuseTriple : reusedUploads) {
            int reusedUploadId = reuseTriple.reused_upload_fk;
            int reusedGroupId = reuseTriple.reused_group_fk;
            int reuseMode = reuseTriple.reuse_mode;
            ItemTreeBounds itemTreeBoundsReused; // = uploadDao->getParentItemBounds(reusedUploadId);
            bool valid = true; // stub
            if (!valid) continue;
            if (reuseMode & UploadDaoConstants::REUSE_ENHANCED) {
                processEnhancedUploadReuse(itemTreeBounds, itemTreeBoundsReused, reusedGroupId);
            } else {
                processUploadReuse(itemTreeBounds, itemTreeBoundsReused, reusedGroupId);
            }
            if (reuseMode & UploadDaoConstants::REUSE_MAIN) {
                reuseMainLicense(uploadId, /*groupId*/0, reusedUploadId, reusedGroupId);
            }
            // TODO: Implement reuseConfSettings if needed
            // if (reuseMode & UploadDaoConstants::REUSE_CONF) {
            //     reuseConfSettings(uploadId, reusedUploadId);
            // }
            if (reuseMode & UploadDaoConstants::REUSE_COPYRIGHT) {
                reuseCopyrights(uploadId, reusedUploadId);
            }
        }
        logInfo("Completed processUploadId for upload " + std::to_string(uploadId));
        return true;
    }
    // --- End ported logic ---

    // Stub methods for demonstration
    std::vector<ReuseTriple> getReusedUploads(int uploadId) {
        // TODO: Replace with real DB/DAO logic
        return {};
    }
    void processEnhancedUploadReuse(const ItemTreeBounds&, const ItemTreeBounds&, int reusedGroupId) {
        logInfo("processEnhancedUploadReuse(" + std::to_string(reusedGroupId) + ")");
        // ...existing code...
    }
    // ...existing code for processUploadReuse (keep only one definition)...
    // Stubs for DB logic
    bool insertReportConfReuse(int uploadId, int reusedUploadId) {
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "INSERT INTO report_conf_reuse (upload_id, reused_upload_id) VALUES (" +
                                std::to_string(uploadId) + ", " +
                                std::to_string(reusedUploadId) + ")";
            txn.exec(query);
            txn.commit();
            return true;
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
            return false;
        }
    }
    void heartbeat(int) {
        // Simulate heartbeat
    }

    void reuseMainLicense(int uploadId, int groupId, int reusedUploadId, int reusedGroupId) {
        // Ported from PHP: reuseMainLicense
        std::cout << "Reusing main license from upload " << reusedUploadId << " to " << uploadId << std::endl;
        std::vector<int> mainLicenseIds = getMainLicenseIds(reusedUploadId, reusedGroupId);
        if (!mainLicenseIds.empty()) {
            std::vector<int> existingMainLicenseIds = getMainLicenseIds(uploadId, groupId);
            for (int mainLicenseId : mainLicenseIds) {
                if (std::find(existingMainLicenseIds.begin(), existingMainLicenseIds.end(), mainLicenseId) != existingMainLicenseIds.end()) {
                    continue;
                } else {
                    makeMainLicense(uploadId, groupId, mainLicenseId);
                }
            }
        }
    }

    // Stubs for DB logic
    std::vector<int> getMainLicenseIds(int uploadId, int groupId) {
        std::vector<int> ids;
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "SELECT license_id FROM main_licenses WHERE upload_id=" +
                                std::to_string(uploadId) + " AND group_id=" + std::to_string(groupId);
            pqxx::result r = txn.exec(query);
            for (const auto &row : r) {
                ids.push_back(row[0].as<int>());
            }
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
        return ids;
    }
    void makeMainLicense(int uploadId, int groupId, int mainLicenseId) {
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "INSERT INTO main_licenses (upload_id, group_id, license_id) VALUES (" +
                                std::to_string(uploadId) + ", " +
                                std::to_string(groupId) + ", " +
                                std::to_string(mainLicenseId) + ")";
            txn.exec(query);
            txn.commit();
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
    }

    void reuseCopyrights(int uploadId, int reusedUploadId) {
        // Ported from PHP: reuseCopyrights
        std::cout << "Reusing copyrights from upload " << reusedUploadId << " to " << uploadId << std::endl;
        int agentId = getAgentId(uploadId);
        int reusedAgentId = getAgentId(reusedUploadId);
        if (agentId == -1 || reusedAgentId == -1) {
            return;
        }
        std::string uploadTreeTableName = getUploadtreeTableName(uploadId);
        std::string extrawhere = " agent_fk=" + std::to_string(agentId);
        auto allCopyrights = getScannerEntries("copyright", uploadTreeTableName, uploadId, extrawhere);
        auto reusedCopyrights = getAllEventEntriesForUpload(reusedUploadId, reusedAgentId);
        if (!reusedCopyrights.empty() && !allCopyrights.empty()) {
            // Pre-index allCopyrights by hash
            std::map<std::string, std::vector<size_t>> copyrightsByHash;
            for (size_t i = 0; i < allCopyrights.size(); ++i) {
                const auto& copyright = allCopyrights[i];
                std::string hash = copyright.hash;
                copyrightsByHash[hash].push_back(i);
            }
            for (const auto& reusedCopyright : reusedCopyrights) {
                std::string reusedHash = reusedCopyright.hash;
                if (copyrightsByHash.count(reusedHash) && !copyrightsByHash[reusedHash].empty()) {
                    size_t copyrightKey = copyrightsByHash[reusedHash].front();
                    copyrightsByHash[reusedHash].erase(copyrightsByHash[reusedHash].begin());
                    const auto& copyright = allCopyrights[copyrightKey];
                    std::string action, content;
                    if (reusedCopyright.is_enabled) {
                        action = "update";
                        content = reusedCopyright.contentedited;
                    } else {
                        action = "delete";
                        content = "";
                    }
                    std::string hash = copyright.hash;
                    auto item = getItemTreeBounds(std::stoi(copyright.uploadtree_pk), uploadTreeTableName);
                    updateTable(item, hash, content, /*userId*/0, "copyright", action);
                    heartbeat(1);
                }
            }
        }
    }

    // Stubs for DB logic and types
    struct CopyrightEntry {
        std::string hash;
        std::string uploadtree_pk;
        bool is_enabled;
        std::string contentedited;
    };
    int getAgentId(int uploadId) {
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "SELECT agent_id FROM agents WHERE upload_id=" + std::to_string(uploadId) + " LIMIT 1";
            pqxx::result r = txn.exec(query);
            if (!r.empty()) {
                return r[0][0].as<int>();
            }
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
        return -1;
    }
    std::string getUploadtreeTableName(int uploadId) {
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "SELECT uploadtree_table FROM uploads WHERE upload_id=" + std::to_string(uploadId) + " LIMIT 1";
            pqxx::result r = txn.exec(query);
            if (!r.empty()) {
                return r[0][0].as<std::string>();
            }
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
        return {};
    }
    std::vector<CopyrightEntry> getScannerEntries(const std::string& agent, const std::string& tableName, int uploadId, const std::string& extrawhere) {
        std::vector<CopyrightEntry> entries;
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "SELECT hash, uploadtree_pk, is_enabled, contentedited FROM " + tableName + " WHERE upload_fk=" + std::to_string(uploadId) + " AND " + extrawhere;
            pqxx::result r = txn.exec(query);
            for (const auto& row : r) {
                CopyrightEntry entry;
                entry.hash = row[0].as<std::string>();
                entry.uploadtree_pk = row[1].as<std::string>();
                entry.is_enabled = row[2].as<bool>();
                entry.contentedited = row[3].as<std::string>();
                entries.push_back(entry);
            }
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
        return entries;
    }
    std::vector<CopyrightEntry> getAllEventEntriesForUpload(int uploadId, int agentId) {
        std::vector<CopyrightEntry> entries;
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string query = "SELECT hash, uploadtree_pk, is_enabled, contentedited FROM copyright_events WHERE upload_id=" + std::to_string(uploadId) + " AND agent_fk=" + std::to_string(agentId);
            pqxx::result r = txn.exec(query);
            for (const auto& row : r) {
                CopyrightEntry entry;
                entry.hash = row[0].as<std::string>();
                entry.uploadtree_pk = row[1].as<std::string>();
                entry.is_enabled = row[2].as<bool>();
                entry.contentedited = row[3].as<std::string>();
                entries.push_back(entry);
            }
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
        return entries;
    }
    ItemTreeBounds getItemTreeBounds(int, const std::string&) { return {}; }
    void updateTable(const ItemTreeBounds&, const std::string& hash, const std::string& content, int userId, const std::string& agent, const std::string& action) {
        try {
            pqxx::connection c("dbname=fossology user=youruser password=yourpass host=localhost");
            pqxx::work txn(c);
            std::string query;
            if (action == "update") {
                query = "UPDATE copyright_table SET contentedited='" + content + "' WHERE hash='" + hash + "' AND user_id=" + std::to_string(userId);
            } else if (action == "delete") {
                query = "DELETE FROM copyright_table WHERE hash='" + hash + "' AND user_id=" + std::to_string(userId);
            }
            txn.exec(query);
            txn.commit();
        } catch (const std::exception &e) {
            std::cerr << "DB error: " << e.what() << std::endl;
        }
    }

    struct ClearingDecision {
        int clearingId;
        int fileId;
        // Add other fields as needed
    };
    std::map<int, ClearingDecision> mapByClearingId(const std::vector<ClearingDecision>& clearingDecisions) {
        std::map<int, ClearingDecision> mapped;
        for (const auto& cd : clearingDecisions) {
            mapped[cd.clearingId] = cd;
        }
        return mapped;
    }
    std::map<int, ClearingDecision> mapByFileId(const std::vector<ClearingDecision>& clearingDecisions) {
        std::map<int, ClearingDecision> mapped;
        for (const auto& cd : clearingDecisions) {
            mapped[cd.fileId] = cd;
        }
        return mapped;
    }
    void processUploadReuse(const ItemTreeBounds&, const ItemTreeBounds&, int reusedGroupId) {
        std::cout << "[Stub] processUploadReuse(" << reusedGroupId << ")" << std::endl;
        // Minimal port of PHP logic
        std::vector<ClearingDecision> clearingDecisions = getFileClearingsFolder(/*itemTreeBoundsReused*/{}, reusedGroupId);
        std::vector<ClearingDecision> currentlyVisibleClearingDecisions = getFileClearingsFolder(/*itemTreeBounds*/{}, /*groupId*/0);
        auto currentlyVisibleById = mapByClearingId(currentlyVisibleClearingDecisions);
        auto clearingDecisionsById = mapByClearingId(clearingDecisions);
        std::vector<ClearingDecision> clearingDecisionsToImport;
        for (const auto& [id, cd] : clearingDecisionsById) {
            if (currentlyVisibleById.find(id) == currentlyVisibleById.end()) {
                clearingDecisionsToImport.push_back(cd);
            }
        }
        auto clearingDecisionToImportByFileId = mapByFileId(clearingDecisionsToImport);
        // Simulate contained items
        std::vector<int> fileIds;
        for (const auto& [fileId, _] : clearingDecisionToImportByFileId) fileIds.push_back(fileId);
        std::vector<Item> containedItems = getContainedItems(fileIds);
        for (const auto& item : containedItems) {
            int fileId = item.fileId;
            if (clearingDecisionToImportByFileId.count(fileId)) {
                createCopyOfClearingDecision(item.id, /*userId*/0, /*groupId*/0, clearingDecisionToImportByFileId[fileId]);
            } else {
                throw std::runtime_error("bad internal state");
            }
            heartbeat(1);
        }
    }
    // Stubs for logic
    std::vector<ClearingDecision> getFileClearingsFolder(const ItemTreeBounds&, int) { return {}; }
    struct Item { int id; int fileId; };
    std::vector<Item> getContainedItems(const std::vector<int>& fileIds) {
        std::vector<Item> items;
        try {
            pqxx::connection c(getDbConnStr());
            pqxx::work txn(c);
            std::string ids = "{";
            for (size_t i = 0; i < fileIds.size(); ++i) {
                ids += std::to_string(fileIds[i]);
                if (i + 1 < fileIds.size()) ids += ",";
            }
            ids += "}";
            std::string query = "SELECT id, file_id FROM contained_items WHERE file_id = ANY('" + ids + "')";
            pqxx::result r = txn.exec(query);
            for (const auto& row : r) {
                Item item;
                item.id = row[0].as<int>();
                item.fileId = row[1].as<int>();
                items.push_back(item);
            }
        } catch (const std::exception &e) {
            logError(std::string("DB error: ") + e.what());
        }
        return items;
    }
    void createCopyOfClearingDecision(int, int, int, const ClearingDecision&) {}
};

int main(int argc, char* argv[]) {
    std::cerr << "[INFO] Reuser agent (C++) starting..." << std::endl;
    if (argc < 2) {
        std::cerr << "Usage: reuser_agent <uploadId>" << std::endl;
        return 1;
    }
    int uploadId = std::stoi(argv[1]);
    ReuserAgent agent;
    bool result = agent.processUploadId(uploadId);
    return result ? 0 : 1;
}
